src/gt4py/stencil_object.py

```diff
@@ -352,48 +433,14 @@ def _call_run(
             ValueError
                 If invalid data or inconsistent options are specified.
         """
-
         if exec_info is not None:
             exec_info["call_run_start_time"] = time.perf_counter()
 
-        domain_ndim = self.domain_info.ndim
-        origin = self._make_origin_dict(origin)
-        all_origin = origin.get("_all_", None)
-
-        # Set an appropriate origin for all fields
-        for name, field_info in self.field_info.items():
-            if field_info is not None:
-                assert name in field_args, f"Missing value for '{name}' field."
-                field_origin = origin.get(name, None)
+        origin = self._normalize_origins(field_args, origin)
```

Comment:
I have two questions here:

1. This assert has been removed. Can you explain why? Is it not needed?
```
        assert (
            len(domain) == domain_ndim
        ), f"Provided domain '{domain}' is not {domain_ndim}-dimensional."
```

2. The code in `_normalize_origins()` was previously inlined here instead of declared in a different function to avoid an extra Python function call at run-time in this performance-critical part of the code. It could have been overkill, but the truth is that the code is not used in any other part anyway, so moving it to a different function does not provide any reusability (at least for now

Reply:
1. The check is the second one in `self._validate_args`, so I didn't feel like it was needed here. I could be convinced otherwise if you feel that it is needed.
2. That was a good thing to do, but I'd like to try it out this way and see if there's a significant performance impact. I think the readability that is gained in `_call_run`, together with the ability to re-use the origin normalization call when using `FrozenStencil` could outweigh the performance impact. For example, although ugly, my test for the `FrozenStencil` so far has been
```python
field_args = {"field_in": in_storage, "field_out": out_storage}
origin = definition._normalize_origins(field_args, None)
domain = definition._get_max_domain(field_args, origin)
frozen = definition.freeze(origin, domain)
exec_info = {"__aggregate_data": True}
for i in range(num_iter):
    frozen(**field_args, exec_info=exec_info)
```

I much prefer the cached version to frozen for this reason.

Reply:
Regarding the point 1. (the removed `assert`), I'm fine to get it removed, but note that it was not redundant since `validate_args` is not always called, and in that case it would have provided more meaningful error messages

Reply:
Regarding the second point, I'm fine with factorizing out the code in `_normalize_origins()` (although it good be good if you can run a couple of microbenchmarks with `timeit` to see if there is a performance regression).

I don't see the use-case to make the methods you mention public. For testing, it shouldn't be a problem to call these methods and if we make them public, it could be a bit confusing for users to expose this functionality as part of the intended interface of `StencilObject`.

Reply:
> (although it good be good if you can run a couple of microbenchmarks with timeit to see if there is a performance regression

Seems to have no effect. I picked a single-statement short test stencil I already wrote and wrapped it with timeit:
```python
import timeit
from gt4py import gtscript
import numpy as np


@gtscript.stencil(backend="gtc:numpy", rebuild=False)
def stencil(
    field_in: gtscript.Field[gtscript.IJK, (np.float_, (2,))],
    field_out: gtscript.Field[gtscript.IJK, np.float_],
):
    with computation(PARALLEL), interval(...):
        field_out = field_in[0, 0, 0][0]


setup = """
from __main__ import stencil
from gt4py import storage as gt_storage
import numpy as np

in_storage = gt_storage.ones(
    shape=(4, 4, 2), backend="gtc:numpy", default_origin=(0, 0, 0), dtype=(np.float_, (2,))
)
out_storage = gt_storage.ones(
    shape=(4, 4, 2),
    backend="gtc:numpy",
    default_origin=(0, 0, 0),
    dtype=np.float_,
)
"""

print(
    timeit.timeit(
        setup=setup,
        stmt="stencil(in_storage, out_storage)",
        number=10000,
    )
)
```

On master this returns ~3.78, and on the `frozen-stencil` branch it's actually a hair faster: 3.70.

I found out that `timeit` isn't super compatible with the way we read the AST in gtscript, so I had to keep that outside the setup code above. I pasted it in case others come across this issue in the future.

Reply:
:+1:

Reply:
> Regarding the point 1. (the removed assert), I'm fine to get it removed, but note that it was not redundant since validate_args is not always called, and in that case it would have provided more meaningful error messages

Definitely. In this sense I agree it was not redundant. However, if I'm a user calling a stencil with `validate_args=False`, I probably don't want that check slowing it down.